src/main/java/io/github/opencubicchunks/cubicchunks/core/server/VanillaNetworkHandler.java

```diff
@@ -210,47 +212,59 @@ public void sendBlockChanges(TShortList dirtyBlocks, Cube cube, EntityPlayerMP p
                             localToBlock(cube.getZ(), z)));
         }
         ((ISPacketMultiBlockChange) packet).setChangedBlocks(updates);
-        ((ISPacketMultiBlockChange) packet).setChunkPos(cube.getCoords().chunkPos());
+        ((ISPacketMultiBlockChange) packet).setChunkPos(new ChunkPos(idx, idz));
 
         player.connection.sendPacket(packet);
     }
 
-    public void updatePlayerPosition(PlayerCubeMap cubeMap, EntityPlayerMP player, int managedPosY) {
+    public void updatePlayerPosition(PlayerCubeMap cubeMap, EntityPlayerMP player, CubePos managedPos) {
         if (!CubicChunksConfig.allowVanillaClients) {
             return;
         }
-        if (!playerYOffsets.containsKey(player)) {
-            playerYOffsets.put(player, 0);
-        }
-        int yOffset = playerYOffsets.get(player);
-        if (managedPosY + yOffset < 2 || managedPosY + yOffset >= 14) {
-            int newYOffset = 8 - managedPosY;
-            playerYOffsets.put(player, newYOffset);
-            switchPlayerOffset(cubeMap, player, yOffset, newYOffset);
+        CubePos offset = playerOffsets.getOrDefault(player, CubePos.ZERO);
+        int idx = managedPos.getX() + offset.getX();
+        int idy = managedPos.getY() + offset.getY();
+        int idz = managedPos.getZ() + offset.getZ();
+
+        boolean shouldSliceTransition = idy < 2 || idy >= 14;
+        boolean isHorizontalSlices = CubicChunksConfig.vanillaClients.horizontalSlices;
+        if (!shouldSliceTransition && isHorizontalSlices
+            && (!CubicChunksConfig.vanillaClients.horizontalSlicesBedrockOnly || ((IBedrockPlayer) player).isBedrock())) {
+            int horizontalSliceSize = CubicChunksConfig.vanillaClients.horizontalSliceSize;
+            int maxHorizontalOffset = Math.max(Math.abs(idx), Math.abs(idz));
+            shouldSliceTransition = maxHorizontalOffset >= Coords.blockToCube(horizontalSliceSize);
+        }
+        if (shouldSliceTransition) {
+            int newXOffset = isHorizontalSlices ? -managedPos.getX() : 0;
+            int newYOffset = 8 - managedPos.getY();
+            int newZOffset = isHorizontalSlices ? -managedPos.getZ() : 0;
```

Comment:
Would be nice to have a somewhat more clever algorithm that avoids putting the player close to the same x/z position as current one, as that would force a re-render. But that may be done separately.

Also this needs testing on 1.14.x+ clients through ViaVersion, as 1.14.x+ is just ignoring chunk packets that are "out of view range" for the client. ViaVersion tells the client that the server has a relatively big (>64 chunks) view radius, but it may not be enough for this purpose.

Reply:
Seeing as how this is only really relevant to Bedrock players, I don't see anyone realistically having any reason to disable `horizontalSlicesBedrockOnly`.

Reply:
Tested the latest commit on 1.15.2 with `horizontalSlicesBedrockOnly` disabled, found no issues.